ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java

```diff
@@ -0,0 +1,211 @@
+ */
+package com.github.ambry.router;
+
+import com.codahale.metrics.Histogram;
+import com.github.ambry.clustermap.PartitionId;
+import com.github.ambry.clustermap.ReplicaId;
+import com.github.ambry.utils.Time;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+
+/**
+ * An implementation of {@link OperationTracker}. It internally maintains the status of a corresponding operation, and
+ * returns information that decides if the operation should continue or terminate.
+ *
+ * This differs from {@link SimpleOperationTracker} in its interpretation of parallelism. A request does not count
+ * towards parallelism if it has been outstanding for more than a configurable cutoff latency (via quantiles) that is
+ * obtained from a {@link Histogram} with latencies of all requests of the same class. In this way it "adapts" to
+ * perceived latencies.
+ */
+class AdaptiveOperationTracker implements OperationTracker {
+  static final long MIN_DATA_POINTS_REQUIRED = 1000;
+
+  private final Time time;
+  private final String datacenterName;
+  private final int successTarget;
+  private final int parallelism;
+  private final double quantile;
+  private final Histogram localColoTracker;
+  private final Histogram crossColoTracker;
+  private final OpTrackerIterator otIterator;
+  private Iterator<ReplicaId> replicaIterator;
+  private final LinkedList<ReplicaId> replicaPool = new LinkedList<ReplicaId>();
+  private final Map<ReplicaId, Long> replicaRequestSendTimes = new HashMap<>();
+
+  private int totalReplicaCount = 0;
+  private int inflightCount = 0;
+  private int succeededCount = 0;
+  private int failedCount = 0;
+
+  private ReplicaId lastReturned = null;
+  private ReplicaId mostRecentlyTracked = null;
+
+  /**
+   * Constructs an {@link AdaptiveOperationTracker}
+   * @param datacenterName The datacenter where the router is located.
+   * @param partitionId The partition on which the operation is performed.
+   * @param crossColoEnabled {@code true} if requests can be sent to remote replicas, {@code false}
+   *                                otherwise.
+   * @param successTarget The number of successful responses required to succeed the operation.
+   * @param parallelism The maximum number of inflight requests at any point of time.
+   * @param time the {@link Time} instance to use.
+   * @param localColoTracker the {@link Histogram} that tracks intra datacenter latencies for this class of requests
+   * @param crossColoTracker the {@link Histogram} that tracks inter datacenter latencies for this class of requests
+   * @param quantile the quantile cutoff to use for when evaluating requests against the trackers.
+   */
+  AdaptiveOperationTracker(String datacenterName, PartitionId partitionId, boolean crossColoEnabled, int successTarget,
```

Comment:
Not a big fan of abstract classes, but perhaps `OperationTracker` could be made one if that makes sense, and these two classes could extend it.